bsi/src/main/java/org/roaringbitmap/bsi/longlong/Roaring64NavigableMapSliceIndex.java

```diff
@@ -0,0 +1,760 @@
+    long value = 0;
+    for (int i = 0; i < this.bitCount(); i += 1) {
+      if (this.bA[i].contains(columnId)) {
+        value |= (1L << i);
+      }
+    }
+
+    return value;
+  }
+
+  /**
+   * RunOptimize attempts to further compress the runs of consecutive values found in the bitmap
+   */
+  public void runOptimize() {
+    this.ebM.runOptimize();
+
+    for (Roaring64NavigableMap integers : this.bA) {
+      integers.runOptimize();
+    }
+    this.runOptimized = true;
+  }
+
+  /**
+   * hasRunCompression returns true if the bitmap benefits from run compression
+   */
+  public boolean hasRunCompression() {
+    return this.runOptimized;
+  }
+
+  /**
+   * GetExistenceBitmap returns a pointer to the underlying existence bitmap of the BSI
+   */
+  public Roaring64NavigableMap getExistenceBitmap() {
+    return this.ebM;
+  }
+
+  /**
+   * the number of slice
+   */
+  public int bitCount() {
+    return this.bA.length;
+  }
+
+  /**
+   * get the cardinality of the bsi (the number of distinct values added to the ebM)
+   */
+  public long getLongCardinality() {
+    return this.ebM.getLongCardinality();
+  }
+
+  /**
+   * get the value at the column ID.
+   */
+  public Pair<Long, Boolean> getValue(long columnId) {
+    boolean exists = this.ebM.contains(columnId);
+    if (!exists) {
+      return Pair.newPair(0L, false);
+    }
+
+    return Pair.newPair(valueAt(columnId), true);
+  }
+
+  private void clear() {
+    this.maxValue = 0;
+    this.minValue = 0;
+    this.ebM = null;
+    this.bA = null;
+  }
+
+  /**
+   * Serialize this bitmapsliceindex(bsi).
+   * <p>
+   * <b>NOTICE:</b> This serialization format is <em>NOT</em> compatible with bsi of roaring(Go)</br>
+   * Current serialization format:<br/>
+   * ---
+   * <pre>
+   * <b>minValue</b> (long with zero-compressed encoding)
+   * <b>maxValue</b> (long with zero-compressed encoding)
+   * <b>runOptimized</b> (boolean)
+   * <b>serializationMode</b> (int with zero-compressed encoding)
+   * <b>ebM</b> as standard Roaring64NavigableMap format (with specify legacy/portable format)
+   * <b>number of slices</b> (int with zero-compressed encoding)
+   * <b>loop over bA[]</b>  as standard Roaring64NavigableMap format (with specify legacy/portable format)
+   * </pre>
+   * ---
+   * <br/>
+   * Consider calling {@link #runOptimize} before serialization to improve compression.
+   * The current bsi is not modified.
+   * </p>
+   * @see <a href="https://github.com/RoaringBitmap/roaring/blob/b32ae1a8cc386adbf7753251b8a580ed37f270a9/roaring64/bsi64.go#L882">bsi64.WriteTo</a>
+   * @param output the DataOutput stream
+   * @throws IOException Signals that an I/O exception has occurred.
+   */
+  public void serialize(DataOutput output) throws IOException {
+    // write meta
+    WritableUtils.writeVLong(output, minValue);
+    WritableUtils.writeVLong(output, maxValue);
+    output.writeBoolean(this.runOptimized);
+    WritableUtils.writeVInt(output, serializationMode);
```

Comment:
You may have a bsiSerializationMode, dedicated to tweaking the BSI serialization. Hence, you may have a bsiImplementationMode compatible with current Go implementation. 

And possibly another implementationMode for what you feel should be the format (for both Java and go). (Possibly as you may also just wait a bit before introducing a second format).

Reply:
do we need three different dimensions of xxxSerializationMode settings here?
users may get confused here